semgrep-core/src/engine/Match_search_mode.ml

```diff
@@ -322,41 +322,99 @@ let run_selector_on_ranges env selector_opt ranges =
+    let fm_mval_range_locs =
+      fm_mvals
+      |> List.filter_map (fun (focus_mvar, mval) ->
+             let* range_loc =
+               Visitor_AST.range_of_any_opt (MV.mvalue_to_any mval)
+             in
+             Some (focus_mvar, mval, range_loc))
     in
-    let focus_range = RM.match_result_to_range focus_match in
-    (* Essentially, we are intersecting the ranges. *)
-    if Range.( $<=$ ) focus_range.r range.r then Some focus_range
-    else if Range.( $<=$ ) range.r focus_range.r then Some range
-    else None
+    let focus_matches =
+      fm_mval_range_locs
+      |> Common.map (fun (focus_mvar, mval, range_loc) ->
+             {
+               PM.rule_id = fake_rule_id (-1, focus_mvar);
+               PM.file = env.xtarget.file;
+               PM.range_loc;
+               PM.tokens = lazy (MV.ii_of_mval mval);
+               PM.env = range.mvars;
+               PM.taint_trace = None;
+             })
+    in
+    let focused_ranges =
+      List.filter_map
+        (fun fms -> intersect (RM.match_result_to_range fms) range)
+        focus_matches
+    in
+    focused_ranges
   in
-  let apply_focus_mvars init_range =
-    focus_mvars
-    |> List.fold_left
-         (fun opt_range (_tok, focus_mvar) ->
-           let* range = opt_range in
-           let final = apply_focus_mvar range focus_mvar in
-           final)
-         (Some init_range)
+  let apply_focus_mvars_list init_range =
+    (* focus_mvars_list and focus_ranges_list are lists of focus statements under a single `patterns`,
+     * ie: [focus-metavariable: $A, $B; focus-metavariable: $X, $Y]
+     *)
+    let focused_ranges_list =
+      focus_mvars_list
+      |> Common.map (fun (_tok, focus_mvars) ->
+             (* focus_mvars is [$A, $B] in a single "focus-metavariable: $A, $B" *)
+             let focused_ranges = apply_focus_mvars focus_mvars init_range in
+             focused_ranges)
+    in
+    let intersect_ranges (list1 : RM.t stack) (list2 : RM.t stack) : RM.t stack
+        =
+      match (list1, list2) with
+      | [ range1 ], [ range2 ] -> (
+          match intersect range1 range2 with
+          | Some r -> [ r ]
+          | None -> [])
+      | [ _ ], []
+      | [], [ _ ] ->
+          []
+      | [], _
+      | _, [] ->
+          failwith "Cannot have an empty `focus-metavariables`"
+      | _, _ ->
+          failwith
+            "Semgrep currently does not support multiple `focus-metavariable` \
+             statements with multiple metavariables under a single `patterns`."
+    in
+    let rec intersect_ranges_list (l : RM.t stack stack) : RM.t stack =
+      match l with
+      | [] -> failwith "No focus-metavariable statements found."
+      | [ first ] -> first
+      | first :: second :: tail ->
+          (* Intersecting two separate focus-metavariable statements at a time *)
+          intersect_ranges_list ([ intersect_ranges first second ] @ tail)
+    in
+    intersect_ranges_list focused_ranges_list
```

Comment:
This is an interesting approach---why not use a fold?

Reply:
i don't think i was able to figure out how to get the types that the function returns to match up with the init, because iirc fold enforces that those are the same.

Reply:
also figuring out how to do this recursively was fun and felt cool